File: common/src/main/kotlin/net/impleri/playerskills/skills/registry/storage/SkillNbtStorage.kt
package net.impleri.playerskills.skills.registry.storage

import net.impleri.playerskills.PlayerSkills
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.NbtIo
import net.minecraft.nbt.StringTag
import net.minecraft.nbt.Tag
import java.io.File
import java.io.IOException

/**
 * Save data in NBT format
 */
internal class SkillNbtStorage : PersistentStorage {
  override fun read(file: File): List<String> {
    val skills: List<String> = ArrayList()

    val tag: CompoundTag = try {
      NbtIo.readCompressed(file)
    } catch (e: IOException) {
      PlayerSkills.LOGGER.debug("Player data file ${file.path} does not exist")
      return skills
    }

    if (!tag.contains(SKILLS_TAG)) {
      PlayerSkills.LOGGER.error("Player data file ${file.path} does not match the expected format")
      return skills
    }

    val list = tag.getList(SKILLS_TAG, Tag.TAG_STRING.toInt())
    return list.stream().map { obj: Tag -> obj.asString }.toList()
  }

  override fun write(file: File, skills: List<String>) {
    val skillList = ListTag()

    skills.stream()
      .map { string -> StringTag.valueOf(string) }
      .forEach { tag -> skillList.add(tag) }

    val storage = CompoundTag()
    storage.put(SKILLS_TAG, skillList)

    try {
      NbtIo.writeCompressed(storage, file)
    } catch (e: IOException) {
      PlayerSkills.LOGGER.error("Failed to write to ${file.path}")
    }
  }

  companion object {
    private const val SKILLS_TAG = "acquiredSkills"
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/registry/storage/SkillNbtStorage.scala
package net.impleri.playerskills.skills.registry.storage

import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.StringTag
import net.minecraft.nbt.Tag

import java.io.File
import java.util.{List => JavaList}
import scala.jdk.javaapi.CollectionConverters
import scala.util.chaining._

trait ReadNbtSkills {
  protected val mcNbt: MinecraftNbt
  protected def readFile(file: File): Either[SkillFileMissing, CompoundTag] =
    mcNbt.read(file)
      .left
      .map(_ => SkillFileMissing(file))

  protected def getList(tag: CompoundTag): Either[SkillFileHasNoData, ListTag] =
    if (tag.contains(SkillNbtStorage.SKILLS_TAG)) Right(tag.getList(SkillNbtStorage.SKILLS_TAG, Tag.TAG_STRING.toInt))
    else Left(SkillFileHasNoData())
}

trait WriteNbtSkills {
  protected val mcNbt: MinecraftNbt

  protected def convertSkills(skills: List[String]): JavaList[StringTag] =
    skills.map(StringTag.valueOf)
      .pipe(CollectionConverters.asJava(_))

  protected def createSkillList(skillsList: JavaList[StringTag]): ListTag =
    new ListTag()
      .tap(_.addAll(skillsList))

  protected def createCompoundTag(skillListTag: ListTag): CompoundTag =
    new CompoundTag()
      .tap(_.put(SkillNbtStorage.SKILLS_TAG, skillListTag))

  protected def tryWrite(file: File)(tag: CompoundTag): Either[NbtFileWriteError, Boolean] =
    mcNbt.write(file, tag)
      .map(_ => true)
      .left.map(_ => FailedToWrite(file))
}

/**
 * Save data in NBT format
 */
class SkillNbtStorage(override val mcNbt: MinecraftNbt) extends PersistentStorage with ReadNbtSkills with WriteNbtSkills {
  override def read(file: File): Either[NbtFileReadError, List[String]] =
    for {
      rawFile <- readFile(file)
      skills <- getList(rawFile)
    } yield CollectionConverters.asScala(skills.iterator)
      .toList
      .map(_.getAsString)

  override def write(file: File, skills: List[String]): Either[NbtFileWriteError, Boolean] =
    convertSkills(skills)
      .pipe(createSkillList)
      .pipe(createCompoundTag)
      .pipe(tryWrite(file))
}

object SkillNbtStorage {
  private[storage] val SKILLS_TAG: String = "acquiredSkills"

  protected[storage] def apply(mcNbt:MinecraftNbt = MinecraftNbtImpl) = new SkillNbtStorage(mcNbt)
}
